### ISSUE TITLE
should Chapel allow implicit conversion from int to uint?

### ISSUE THREAD

**Issue by mppf** [chapel-lang/chapel#20091]
Today, the situation is that if you have e.g. `myInt + myUint` then a special overload in ChapelBase causes that to be an error. E.g. `inline operator +(a: uint(64), b: int(64)) { _throwOpError("+"); }`. But that special overload causes other problems, e.g. #17756. Another issue with that approach is that it makes it harder for somebody to make a custom binary operator that handles all of the numeric types since one has to make `int(?w), uint(w)` and `uint(?w), int(w)` versions to get the same behavior as `+`.

For example, here is a test file that defines `proc f` accepting pairs of each numeric type (`proc f(a: int(32), b: int(32))`, `...`, `proc f(a: real, b: real)` etc). https://github.com/chapel-lang/chapel/blob/417d0d064bfc437baea87ec490b462adabac74d4/test/types/coerce/allNumericsBinary.chpl . The behavior of such an overload with `int` and `uint` actual arguments is not ideal, in my opinion. Here is what happens today:
 * if the arguments are maximum width int/uint, it calls the `real` version (because `int` and `uint` can both implicitly convert to `real`).
 * if the arguments are smaller, e.g. `int(32)`/`uint(32)`, it calls the `int` version of the next size, so `int(8)`/`uint(8)` results in `int(16)` and so on.

I think in both of the above cases, the current behavior can be very surprising to a user who thought that they were working with one numeric type / size of number.

This is a problem that C# has and that I think Chapel inherited by drawing on some details of C# in terms of which numeric implicit conversions are allowed. In contrast, C++ does not have this problem, because it allows almost all implicit numeric conversions.

PR #19959 investigates adding a special heuristic filter on candidates before the disambiguation process to address this problem. But, that heuristic filter is a heuristic and so it isn't as essential or principled as it could be. This issue is proposing an alternative strategy.

What if we simply allowed implicit conversion from `int(w)` to `uint(w)`?

It would cause `myInt + myUint` to result in a `uint`. I think that this strategy is defensible because:
 1. `int` can convert to `uint` and back again without any loss of precision
 2. which implicit conversions are allowed isn't really about overflow because `int + int` or `uint + uint` can easily overflow on their own. So I don't think we should be worried about preventing overflow by preventing `int + uint`.
 3. `uint` is a more specialized type than `int` (and in fact it can wrap around without running into undefined behavior). Similarly, I view `real` as a more specialized type than `int`. So it makes sense to me that `int` can convert to `real` and that `int` could convert to `uint`.

A nice side effect of this rule would be that it would make the "prefer int types over uint types" rule in https://chapel-lang.org/docs/language/spec/procedures.html#determining-more-specific-functions more natural. In particular, it becomes part of the usual "prefer a more specific type" rule (i.e., prefer int over uint because int can implicitly convert to uint, but uint cannot implicitly convert to int).

**Comment by mppf** [comment]
There is also this comment from the code introducing the error for `myInt + myUint`:

https://github.com/chapel-lang/chapel/commit/4a08306cb8f6dbfc5db2b0c4c342171f78ea1f35#diff-e97600672e28c3968962b105840b7534d5ebae840773a9b7672d02dc48768800R1159

```
  // non-param/non-param -- these are designed to throw an error
  // stating that these combinations are not supported; it's an
  // open topic of discussion whether we should support them and
  // have the return type default to int or uint; and/or whether
  // this error case should be made a lower-priority fallback,
  // permitting the user to define their own overloads (or 'use'
  // a module which does).
  //
```

**Comment by mppf** [comment]
In C++, the equivalent of `int` + `uint` results in a `uint`.

**Comment by mppf** [comment]
> In C++, the equivalent of `int` + `uint` results in a `uint`.

Here is a demonstrator program:

``` c++
#include <cinttypes>
#include <cstdio>

const char* tname(int8_t x)  { return "int8_t"; }
const char* tname(int16_t x) { return "int16_t"; }
const char* tname(int32_t x) { return "int32_t"; }
const char* tname(int64_t x) { return "int64_t"; }

const char* tname(uint8_t x)  { return "uint8_t"; }
const char* tname(uint16_t x) { return "uint16_t"; }
const char* tname(uint32_t x) { return "uint32_t"; }
const char* tname(uint64_t x) { return "uint64_t"; }

const char* tname(float x)  { return "float"; }
const char* tname(double x) { return "double"; }

int main() {
  int myInt = 0;
  unsigned int myUint = 1;
  printf("myInt + myUint is %s\n", tname(myInt + myUint));

  long myLong = 0;
  unsigned long myULong = 1;
  printf("myLong + myULong is %s\n", tname(myLong + myULong));
}

```

```
$ g++ prog.cpp && ./a.out
myInt + myUint is uint32_t
myLong + myULong is uint64_t
```

**Comment by mppf** [comment]
I'm closing this issue because we have released Chapel 1.28 with the implicit conversions allowed and they are documented. (See PRs #20528 and #20623).
